<a href="https://colab.research.google.com/github/DhananjayaFdo/Databases-and-Analytics-Assignment/blob/main/mongoDB_development.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Connect in Colab

In [5]:
!pip install pymongo

from pymongo import MongoClient
import pandas as pd
from datetime import datetime
import random
from datetime import datetime, timedelta

connection_string = "mongodb+srv://dhananjaya08fdo_db_user:GB6HW6LSDTaYXamO@cluster0.rqymrh2.mongodb.net/"

client = MongoClient(connection_string)

db = client["northstar_db"]

print("Connected to MongoDB Atlas ✓")
print("Database:", db.name)

Connected to MongoDB Atlas ✓
Database: northstar_db


# Load and transform data

In [6]:
baseUrl = "https://raw.githubusercontent.com/DhananjayaFdo/Databases-and-Analytics-Assignment/refs/heads/main/northstar_dataset";

complaints = pd.read_csv (f'{baseUrl}/complaints.csv');
app_events = pd.read_csv (f'{baseUrl}/app_events.csv');

complaints["compensation_amount"] = complaints["compensation_amount"].fillna(0)
app_events["order_id"] = app_events["order_id"].fillna("N/A")

print("Complaints shape:", complaints.shape)
print("App events shape:", app_events.shape)

Complaints shape: (320, 10)
App events shape: (640, 10)


In [7]:
def complaint_to_doc(row):
    return {
        "_id":                  row["complaint_id"],
        "customer_id":          row["customer_id"],
        "order_id":             row["order_id"],
        "complaint_type":       row["complaint_type"],
        "channel":              row["channel"],
        "severity":             row["severity"],
        "created_at":           row["created_at"],
        "status":               row["status"],
        "resolution_days":      row["resolution_days"],
        "compensation_amount":  row["compensation_amount"]
    }

complaint_docs = [complaint_to_doc(row) for _, row in complaints.iterrows()]

print("Sample complaint document:")
print(complaint_docs[0])

Sample complaint document:
{'_id': 'CP0001', 'customer_id': 'C0464', 'order_id': 'O00814', 'complaint_type': 'AppIssue', 'channel': 'App', 'severity': 'High', 'created_at': '2025-03-30 02:36:00', 'status': 'Open', 'resolution_days': 11, 'compensation_amount': 23.99}


In [8]:
session_docs = []

for session_id, group in app_events.groupby("session_id"):
    events_list = []
    for _, row in group.iterrows():
        events_list.append({
            "event_id":      row["event_id"],
            "event_type":    row["event_type"],
            "timestamp":     row["event_timestamp"],
            "order_id":      row["order_id"],
            "latency_ms":    row["api_latency_ms"],
            "success":       bool(row["success_flag"])
        })

    doc = {
        "_id":          session_id,
        "customer_id":  group.iloc[0]["customer_id"],
        "device_type":  group.iloc[0]["device_type"],
        "zone_context": group.iloc[0]["zone_context"],
        "event_count":  len(events_list),
        "events":       events_list
    }
    session_docs.append(doc)

print(f"Total sessions created: {len(session_docs)}")
print("\nSample session document:")
print(session_docs[0])

Total sessions created: 637

Sample session document:
{'_id': 'S10192', 'customer_id': 'C0180', 'device_type': 'Android', 'zone_context': 'Ctr', 'event_count': 1, 'events': [{'event_id': 'AE00403', 'event_type': 'payment_retry', 'timestamp': '2025-03-19 09:20:00', 'order_id': 'O00149', 'latency_ms': 108, 'success': True}]}


"Rather than storing 640 flat event rows, events are grouped into session documents with embedded arrays. This means one read retrieves the full session history — no JOIN required. This is the core advantage of document modelling over relational tables for event data."

# CRUD

## CREATE (insert data)

In [9]:
complaints_col = db["complaints"]
complaints_col.drop()

result = complaints_col.insert_many(complaint_docs)
print(f"Inserted {len(result.inserted_ids)} complaint documents ✓")

Inserted 320 complaint documents ✓


## CREATE RANDOM DATA

In [17]:
complaint_types  = ["Delay", "MissedPickup", "AppIssue", "DriverBehaviour",
                    "SupportExperience", "Billing", "Damage"]
channels         = ["App", "Phone", "Chatbot", "Email"]
severities       = ["Low", "Medium", "High"]
statuses         = ["Open", "Resolved", "Escalated", "AwaitingCustomer"]

type_weights     = [35, 20, 15, 15, 7, 5, 3]
severity_weights = [22, 54, 24]
status_weights   = [17, 58, 12, 13]

def random_date(start_year=2024, end_year=2025):
    start = datetime(start_year, 1, 1)
    end   = datetime(end_year, 12, 31)
    delta = end - start
    return (start + timedelta(days=random.randint(0, delta.days))).strftime("%Y-%m-%d %H:%M:%S")

def generate_complaints(n=50000):
    docs = []
    for i in range(1, n + 1):
        severity = random.choices(severities, weights=severity_weights)[0]
        status   = random.choices(statuses,   weights=status_weights)[0]

        if severity == "High":
            res_days    = random.randint(5, 20)
            comp_amount = round(random.uniform(20, 80), 2)
        elif severity == "Medium":
            res_days    = random.randint(2, 10)
            comp_amount = round(random.uniform(5, 30), 2)
        else:
            res_days    = random.randint(1, 5)
            comp_amount = round(random.uniform(0, 15), 2)

        if status == "Escalated":
            comp_amount = max(comp_amount, 15.0)

        docs.append({
            "_id":                 f"GEN{i:05d}",
            "customer_id":         f"C{random.randint(1, 650):04d}",
            "order_id":            f"O{random.randint(1, 1250):05d}",
            "complaint_type":      random.choices(complaint_types, weights=type_weights)[0],
            "channel":             random.choice(channels),
            "severity":            severity,
            "created_at":          random_date(),
            "status":              status,
            "resolution_days":     res_days,
            "compensation_amount": comp_amount
        })
    return docs

# Generate
generated_docs = generate_complaints(50000)
print(f"Generated {len(generated_docs)} documents ✓")

complaints_col.drop()
complaints_col.insert_many(generated_docs)

print(f"Inserted {len(result.inserted_ids)} documents into MongoDB ✓")

Generated 50000 documents ✓
Inserted 3000 documents into MongoDB ✓


In [7]:
sessions_col = db["app_sessions"]
sessions_col.drop()

result = sessions_col.insert_many(session_docs)
print(f"Inserted {len(result.inserted_ids)} session documents ✓")

Inserted 637 session documents ✓


## READ (find and filter)

In [8]:
high_severity = list(complaints_col.find(
    {"severity": "High"},
    {"_id": 1, "customer_id": 1, "complaint_type": 1, "status": 1}
))

print(f"High severity complaints: {len(high_severity)}")
for doc in high_severity[:3]:
    print(doc)

High severity complaints: 77
{'_id': 'CP0001', 'customer_id': 'C0464', 'complaint_type': 'AppIssue', 'status': 'Open'}
{'_id': 'CP0003', 'customer_id': 'C0469', 'complaint_type': 'Delay', 'status': 'Open'}
{'_id': 'CP0008', 'customer_id': 'C0309', 'complaint_type': 'AppIssue', 'status': 'Resolved'}


In [9]:
escalated = list(complaints_col.find({"status": "Escalated"}))
print(f"Escalated complaints: {len(escalated)}")

Escalated complaints: 38


In [10]:
busy_sessions = list(sessions_col.find(
    {"event_count": {"$gt": 3}},
    {"_id": 1, "customer_id": 1, "event_count": 1, "device_type": 1}
))

print(f"Sessions with more than 3 events: {len(busy_sessions)}")
for s in busy_sessions[:3]:
    print(s)

Sessions with more than 3 events: 0


In [11]:
payment_retry_sessions = list(sessions_col.find(
    {"events.event_type": "payment_retry"},
    {"_id": 1, "customer_id": 1, "event_count": 1}
))

print(f"Sessions with payment retry: {len(payment_retry_sessions)}")

Sessions with payment retry: 69


"MongoDB allows querying inside nested arrays — 'events.event_type' searches within embedded event lists without any joins. This shows the operational advantage of the session document design."

## UPDATE

In [12]:
update_result = complaints_col.update_one(
    {"_id": "CP0001"},
    {"$set": {"status": "Resolved", "resolution_days": 5}}
)

print("Matched:", update_result.matched_count)
print("Modified:", update_result.modified_count)

print(complaints_col.find_one({"_id": "CP0001"}))

Matched: 1
Modified: 1
{'_id': 'CP0001', 'customer_id': 'C0464', 'order_id': 'O00814', 'complaint_type': 'AppIssue', 'channel': 'App', 'severity': 'High', 'created_at': '2025-03-30 02:36:00', 'status': 'Resolved', 'resolution_days': 5, 'compensation_amount': 23.99}


In [13]:
bulk_result = complaints_col.update_many(
    {"complaint_type": "Delay"},
    {"$set": {"priority_flag": True}}
)

print(f"Updated {bulk_result.modified_count} Delay complaints with priority flag ✓")

Updated 101 Delay complaints with priority flag ✓


## DELETE

In [14]:
complaints_col.insert_one({
    "_id": "TEST001",
    "complaint_type": "Test",
    "status": "Test"
})

delete_result = complaints_col.delete_one({"_id": "TEST001"})
print("Deleted:", delete_result.deleted_count, "document ✓")

Deleted: 1 document ✓


# AGGREGATION pipeline

In [15]:
pipeline = [
    {
        "$group": {
            "_id": {
                "type":     "$complaint_type",
                "severity": "$severity"
            },
            "total":              {"$sum": 1},
            "avg_resolution":     {"$avg": "$resolution_days"},
            "total_compensation": {"$sum": "$compensation_amount"}
        }
    },
    {"$sort": {"total": -1}},
    {"$limit": 10}
]

results = list(complaints_col.aggregate(pipeline))
for r in results:
    print(r)

{'_id': {'type': 'Delay', 'severity': 'Medium'}, 'total': 56, 'avg_resolution': 5.964285714285714, 'total_compensation': 964.87}
{'_id': {'type': 'MissedPickup', 'severity': 'Medium'}, 'total': 37, 'avg_resolution': 6.162162162162162, 'total_compensation': 644.8}
{'_id': {'type': 'DriverBehaviour', 'severity': 'Medium'}, 'total': 31, 'avg_resolution': 5.419354838709677, 'total_compensation': 476.29}
{'_id': {'type': 'Delay', 'severity': 'Low'}, 'total': 27, 'avg_resolution': 6.481481481481482, 'total_compensation': 220.43}
{'_id': {'type': 'AppIssue', 'severity': 'Medium'}, 'total': 25, 'avg_resolution': 7.36, 'total_compensation': 386.58}
{'_id': {'type': 'Delay', 'severity': 'High'}, 'total': 18, 'avg_resolution': 12.444444444444445, 'total_compensation': 511.54}
{'_id': {'type': 'DriverBehaviour', 'severity': 'High'}, 'total': 16, 'avg_resolution': 13.75, 'total_compensation': 460.63}
{'_id': {'type': 'MissedPickup', 'severity': 'High'}, 'total': 16, 'avg_resolution': 11.5625, 'tota

In [16]:
engagement_pipeline = [
    {
        "$group": {
            "_id":          "$customer_id",
            "total_events": {"$sum": "$event_count"},
            "sessions":     {"$sum": 1}
        }
    },
    {"$sort": {"total_events": -1}},
    {"$limit": 10}
]

top_users = list(sessions_col.aggregate(engagement_pipeline))
print("Most active customers:")
for u in top_users:
    print(u)

Most active customers:
{'_id': 'C0540', 'total_events': 5, 'sessions': 5}
{'_id': 'C0442', 'total_events': 4, 'sessions': 4}
{'_id': 'C0256', 'total_events': 4, 'sessions': 4}
{'_id': 'C0401', 'total_events': 4, 'sessions': 4}
{'_id': 'C0568', 'total_events': 4, 'sessions': 4}
{'_id': 'C0495', 'total_events': 4, 'sessions': 4}
{'_id': 'C0182', 'total_events': 4, 'sessions': 4}
{'_id': 'C0266', 'total_events': 4, 'sessions': 4}
{'_id': 'C0634', 'total_events': 4, 'sessions': 4}
{'_id': 'C0479', 'total_events': 4, 'sessions': 4}


"The aggregation pipeline processes data server-side inside MongoDB — no need to pull all records into Python first. This is how MongoDB supports scalable analytics on large operational datasets."